src/test/java/com/microsoft/sqlserver/clientcertauth/ClientCertificateAuthenticationTest.java

```diff
@@ -0,0 +1,172 @@
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+    
+    @Test
+    public void pkcs8EncryptedTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + ".pem;" + "clientKey="
+                + clientKey + "-encrypted-pkcs8.key;" + "clientKeyPassword=" + clientKeyPassword + ";";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+
+    @Test
+    public void pfxTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + ".pfx;";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+
+    @Test
+    public void pfxEncrytedTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + "-encrypted.pfx;" + "clientKeyPassword="
+                + clientKeyPassword + ";";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+    
+    @Test
+    public void pvkTest() throws Exception {
+        String conStr = connectionString + ";clientCertificate=" + clientCertificate + ".cer;" + "clientKey="
+                + clientKey + ".pvk;" + "clientKeyPassword=" + clientKeyPassword + ";";
+        try (Connection conn = DriverManager.getConnection(conStr); Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT @@VERSION AS 'SQL Server Version'");
+            rs.next();
+            assertTrue(rs.getString(1).contains(TestResource.getResource("R_microsoft")));
+        }
+    }
+
+    @Test
+    public void invalidCert() throws Exception {
+        String conStr = connectionString + ";clientCertificate=invalid_path;" + "clientKeyPassword=" + clientKeyPassword
+                + ";";
+        try (Connection conn = DriverManager.getConnection(conStr)) {
+        } catch (SQLServerException e) {
+            assertTrue(e.getMessage().contains(TestResource.getResource("R_invalidPath")));

```

Comment:
You can use `R_clientCertError` from SQLServerResource.

Reply:
SQLServerResource is not visible in this package, we can just use the TestResource for test related texts.

Reply:
See SQLServerConnectionTest.executeInvalidFmt() for example use.

`assertTrue(e.getMessage().matches(TestUtils.formatErrorMsg("R_invalidArgument")));`

Reply:
Thanks. done.